audio_service/example/lib/example_android_songs.dart

```diff
@@ -0,0 +1,588 @@
+///    from `MediaStore`
+///  * because [AndroidContentResolver] does not support (yet)
+///    `ContentResolver.openFileDescriptor`
+///
+/// To run this example, use:
+///
+/// flutter run -t lib/example_android_songs.dart
+
+import 'dart:io';
+import 'dart:typed_data';
+
+import 'package:audio_service/audio_service.dart';
+import 'package:flutter/material.dart';
+import 'package:android_content_provider/android_content_provider.dart';
+import 'package:device_info_plus/device_info_plus.dart';
+import 'package:just_audio/just_audio.dart';
+import 'package:rxdart/rxdart.dart';
+import 'package:permission_handler/permission_handler.dart';
+
+// You might want to provide this using dependency injection rather than a
+// global variable.
+late AudioPlayerHandler _audioHandler;
+late int sdkInt;
+List<Song>? _songs;
+final albumArtPaths = <int, String>{};
+
+/// Whether running on scoped storage (Android 10 and above),
+/// and should use bytes to load album arts from `MediaStore`.
+bool get useScopedStorage => sdkInt >= 29;
+
+Future<void> main() async {
+  WidgetsFlutterBinding.ensureInitialized();
+
+  DeviceInfoPlugin deviceInfo = DeviceInfoPlugin();
+  AndroidDeviceInfo androidInfo = await deviceInfo.androidInfo;
+  sdkInt = androidInfo.version.sdkInt!;
+
+  _audioHandler = await AudioService.init(
+    builder: () => AudioPlayerHandler(),
+    config: const AudioServiceConfig(
+      androidNotificationChannelId: 'com.ryanheise.myapp.channel.audio',
+      androidNotificationChannelName: 'Audio playback',
+      androidNotificationOngoing: true,
+    ),
+  );
+  runApp(const MyApp());
+}
+
+class MyApp extends StatelessWidget {
+  const MyApp({Key? key}) : super(key: key);
+
+  @override
+  Widget build(BuildContext context) {
+    return MaterialApp(
+      title: 'Audio Service Demo',
+      theme: ThemeData(primarySwatch: Colors.blue),
+      home: const MainScreen(),
+    );
+  }
+}
+
+class MainScreen extends StatefulWidget {
+  const MainScreen({Key? key}) : super(key: key);
+
+  @override
+  State<MainScreen> createState() => _MainScreenState();
+}
+
+class _MainScreenState extends State<MainScreen> {
+  PermissionStatus? _permissionStatus;
+
+  @override
+  void initState() {
+    super.initState();
+    _fetch();
+  }
+
+  Future<void> _fetch() async {
+    _permissionStatus = await Permission.storage.request();
+    if (mounted) {
+      setState(() {
+        // Update to maybe show progress indicator.
+      });
+    }
+    if (_permissionStatus == PermissionStatus.granted) {
+      await _fetchSongs();
+      await _fetchArts();
+      if (_songs!.isNotEmpty) {
+        _audioHandler.init();
+      }
+    } else if (_permissionStatus == PermissionStatus.permanentlyDenied) {
+      await openAppSettings();
+    }
+    if (mounted) {
+      setState(() {
+        // Update to show the fetch result.
+      });
+    }
+  }
+
+  Future<void> _fetchSongs() async {
+    await autoCloseScope(() async {
```

Comment:
This discussion probably belongs on your plugin, but looking at the source for `autoCloseScope`, I notice that it sets a global state `_inCloseScope`. What happens when you run two `autoCloseScope`s in parallel?

Reply:
Yeah, thanks for spotting this, that was definitely an oversight on my part. I wanted to prevent nested scopes, because it doesn't really make sense to create them, but now I see it wasn't a good idea - will remove this when I have a moment and ensure that it works with a test.